test/e2e/persistent_volumes.go

```diff
@@ -465,78 +429,153 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
 		}
 	})
 
-	// Individual tests follow:
-	//
-	// Create an nfs PV, then a claim that matches the PV, and a pod that
-	// contains the claim. Verify that the PV and PVC bind correctly, and
-	// that the pod can write to the nfs volume.
-	It("should create a non-pre-bound PV and PVC: test write access [Flaky]", func() {
+	Context("with Single PV - PVC pairs", func() {
 
-		pv, pvc, err = createPVPVC(c, serverIP, ns, false)
-		if err != nil {
-			framework.Failf("%v", err)
-		}
+		var pv *api.PersistentVolume
+		var pvc *api.PersistentVolumeClaim
 
-		// validate PV-PVC, create and verify writer pod, delete PVC
-		// and PV
-		pv, pvc, err = completeTest(f, c, ns, pv, pvc)
-		if err != nil {
-			framework.Failf("%v", err)
-		}
+		AfterEach(func() {
+			if c != nil && len(ns) > 0 { // still have client and namespace
+				if pvc != nil && len(pvc.Name) > 0 {
+					// Delete the PersistentVolumeClaim
+					framework.Logf("AfterEach: PVC %v is non-nil, deleting claim", pvc.Name)
+					err := c.PersistentVolumeClaims(ns).Delete(pvc.Name)
+					Expect(err).NotTo(HaveOccurred())
+					pvc = nil
+				}
+				if pv != nil && len(pv.Name) > 0 {
+					// Delete the PersistentVolume
+					framework.Logf("AfterEach: PV %v is non-nil, deleting pv", pv.Name)
+					err := c.PersistentVolumes().Delete(pv.Name)
+					Expect(err).NotTo(HaveOccurred())
+					pv = nil
+				}
+			}
+		})
+
+		// Individual tests follow:
+		//
+		// Create an nfs PV, then a claim that matches the PV, and a pod that
+		// contains the claim. Verify that the PV and PVC bind correctly, and
+		// that the pod can write to the nfs volume.
+		It("should create a non-pre-bound PV and PVC: test write access [Flaky]", func() {
+
+			pv, pvc = createPVPVC(c, serverIP, ns, false)
+			defer deletePersistentVolume(c, pv)
```

Comment:
How does this defer work when, just before the func returns, you nil the pv and pvc?